_posts/2018-09-26-Inter-service-communication.md

```diff
@@ -0,0 +1,216 @@
+---
+layout: post
+authors: [kevin_van_houtte]
+title: 'Inter-service/inter-process communication with Feign: Tips & Tricks'
+image: /img/intercommunication/intercomm_header.jpg
+tags: [Microservices, Security, Spring, Cloud, Feign, OpenFeign]
+category: Microservices
+comments: true
+---
+
+In contrast to monolithic applications, services in a distributed system are running on multiple machines. 
+To let these services interact with each other, we need some kind of inter-process communication mechanism.
+With the help of Feign, I will explain how we can fire off synchronous calls to another service.
+
+
+# Feign
+To understand the basics of inter-process communication, we need to look at what kind of interactions we can do.
+Feign, a declarative HTTP client by Netflix simplifies our way of interacting with other services. 
+When we decide that it is time to decompose our modulith because of numerous reasons, we would have to look for a way to handle our inter-process communication.
+Feign provides support for Spring, load balancing, resilience, security and lots more and I will tell you how they work and how to use them inside your Spring application. 
+
+# Setup
+To use Feign we need to add it to our classpath
+
+{% highlight xml %}
+    <dependency>
+        <groupId>org.springframework.cloud</groupId>
+        <artifactId>spring-cloud-starter-openfeign</artifactId>
+    </dependency>
+{% endhighlight %}
+
+When we inspect the dependency module, we see that there is a lot coming out-of-the-box with the Spring Cloud Starter.
+If you are providing your own resilience or load balancing library you can just add the necessary dependencies you need.
+Be aware that the syntax is different between using the Spring wrapper or OpenFeign itself.
+To let your Spring context know that it has to search for Feign clients, we just add `@EnableFeignClients` on top of the main configuration class. 
+
+# Feign Clients 
+After we've enabled Feign on our classpath, we can start adding Feign clients. 
+When defining our clients, we have two solutions you can choose from. 
+The OpenFeign library, which provides us with the basic but very customizable Feign clients, and the Spring library, that adds a few extra libraries to it for cloud tooling.
+
+## Spring
+
+{% highlight java %}
+    @FeignClient(value = "auth", fallback = FallbackAuthClient.class, configuration = FeignConfig.class)
+    public interface AuthClient { 
+        @RequestMapping(method = RequestMethod.GET, value = "checkToken")
+        boolean isTokenValid(@RequestHeader("Authorization") String token);
+    }
+{% endhighlight %}
+
+Explanation properties:
+
+* `@FeignClient`: is the annotation for Spring to recognize Feign clients, Feign clients have to be interfaces as it is self declarative.
+* `value/name`: is the name of the application you are communicating with, Service Discovery ( when used ) uses this to bind them to the correct IP
```

Comment:
This is not the name of the application, but the name of the Feign client. This will be used to create a Ribbon load balancer which can then be linked to the target application using service discovery or a fixed list of servers.
You could also use the `url` attribute to point your client to the target application when you're not using Ribbon.